ocaml/xapi/pool_periodic_update_sync.ml

```diff
@@ -0,0 +1,193 @@
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = "pool_periodic_update_sync" end)
+
+open D
+open Client
+
+let finally = Xapi_stdext_pervasives.Pervasiveext.finally
+
+let periodic_update_sync_task_name = "Periodic update synchronization"
+
+let update_sync_minimum_interval = 60. *. 60. *. 2.
+
+exception UpdateSync_RetryNumExceeded of int
+
+let seconds_random_within_a_day () =
+  if Xapi_fist.disable_periodic_update_sync_sec_randomness () then
+    0.
+  else
+    let secs_of_a_day = 24. *. 60. *. 60. in
+    let secs_random = Random.float secs_of_a_day in
+    secs_random
+
+let frequency_to_str ~frequency =
+  match frequency with `daily -> "daily" | `weekly -> "weekly"
+
+let weekday_to_int = function
+  | `Sun ->
+      0
+  | `Mon ->
+      1
+  | `Tue ->
+      2
+  | `Wed ->
+      3
+  | `Thu ->
+      4
+  | `Fri ->
+      5
+  | `Sat ->
+      6
+
+let utc_start_of_next_scheduled_day ~utc_now ~tz_offset_s ~frequency
+    ~day_configed_int =
+  let y, m, d = fst (Ptime.to_date_time ~tz_offset_s utc_now) in
+  let utc_start_of_today =
+    Ptime.of_date_time ((y, m, d), ((0, 0, 0), tz_offset_s)) |> Option.get
+  in
+
+  match frequency with
+  | `daily ->
+      (* schedule in the next day *)
+      let one_day_span = Ptime.Span.of_d_ps (1, 0L) |> Option.get in
+      Ptime.add_span utc_start_of_today one_day_span |> Option.get
+  | `weekly ->
+      let days =
+        let today_wday = Ptime.weekday ~tz_offset_s utc_now |> weekday_to_int in
+        match today_wday < day_configed_int with
+        | true ->
+            day_configed_int - today_wday
+        | false ->
+            day_configed_int + 7 - today_wday
+      in
+      let span = Ptime.Span.of_d_ps (days, 0L) |> Option.get in
+      Ptime.add_span utc_start_of_today span |> Option.get
+
+let next_scheduled_datetime ~delay ~utc_now ~tz_offset_s =
+  let span = Ptime.Span.of_float_s delay |> Option.get in
+  let target = Ptime.add_span utc_now span |> Option.get in
+  let (y, m, d), ((hh, mm, ss), _) = Ptime.to_date_time ~tz_offset_s target in
+  (y, m, d, hh, mm, ss)
+
+let print_next_schedule ~delay ~utc_now ~tz_offset_s =
+  debug "[PeriodicUpdateSync] delay for next update sync: %f seconds" delay ;
+  let y, m, d, hh, mm, ss =
+    next_scheduled_datetime ~delay ~utc_now ~tz_offset_s
+  in
+  debug
+    "[PeriodicUpdateSync] next update sync scheduled at pool time: %d-%d-%d, \
+     %d:%d:%d "
+    y m d hh mm ss
```

Comment:
Hi Paul, now there are XenRT test cases designed based on the log: test case F3 in https://info.citrite.net/pages/viewpage.action?spaceKey=xenserver&title=Feature+Quality+Plan+for+CP-41451+Toolstack+part#comment-1635851208, and since this is logged in the lowest level: debug, and it does not spam too much: for daily schedule, it is logged once per day, and for weekly schedule, it is logged once per week, can we leave it also for easy debugging in case of customer's issue?

Reply:
Oh, I forgot that, sorry, thanks for helping removed the internal link!

> Using the logs for testing is very easy to break. We offer guarantees for the API
Yes, you are right.
OK, I will discuss with Qin on the test plan again.